File: nprpc_benchmark/web/nprpc_benchmark/src/rpc/benchmark.ts
import * as NPRPC from 'nprpc'

const u8enc = new TextEncoder();
const u8dec = new TextDecoder();

export interface Data1 {
  f1: string;
  f2: number/*f32*/;
}

export namespace Flat_benchmark {
export class Data1_Direct extends NPRPC.Flat.Flat {
  public get f1() {
    const offset = this.offset + 0;
    const n = this.buffer.dv.getUint32(offset + 4, true);
    return n > 0 ? u8dec.decode(new DataView(this.buffer.array_buffer, offset + this.buffer.dv.getUint32(offset, true), n)) : ""
  }
  public set f1(str: string) {
    const bytes = u8enc.encode(str);
    const offset = NPRPC.Flat._alloc(this.buffer, this.offset + 0, bytes.length, 1, 1);
    new Uint8Array(this.buffer.array_buffer, offset).set(bytes);
  }
  public get f2() { return this.buffer.dv.getFloat32(this.offset+8,true); }
  public set f2(value: number) { this.buffer.dv.setFloat32(this.offset+8,value,true); }
}
} // namespace Flat 
export interface Data {
  f1: string;
  f2: string;
  f3: number/*u8*/;
  f4: number/*u16*/;
  f5: number/*u32*/;
  f6: Array<Data1>;
}

export namespace Flat_benchmark {
export class Data_Direct extends NPRPC.Flat.Flat {
  public get f1() {
    const offset = this.offset + 0;
    const n = this.buffer.dv.getUint32(offset + 4, true);
    return n > 0 ? u8dec.decode(new DataView(this.buffer.array_buffer, offset + this.buffer.dv.getUint32(offset, true), n)) : ""
  }
  public set f1(str: string) {
    const bytes = u8enc.encode(str);
    const offset = NPRPC.Flat._alloc(this.buffer, this.offset + 0, bytes.length, 1, 1);
    new Uint8Array(this.buffer.array_buffer, offset).set(bytes);
  }
  public get f2() {
    const offset = this.offset + 8;
    const n = this.buffer.dv.getUint32(offset + 4, true);
    return n > 0 ? u8dec.decode(new DataView(this.buffer.array_buffer, offset + this.buffer.dv.getUint32(offset, true), n)) : ""
  }
  public set f2(str: string) {
    const bytes = u8enc.encode(str);
    const offset = NPRPC.Flat._alloc(this.buffer, this.offset + 8, bytes.length, 1, 1);
    new Uint8Array(this.buffer.array_buffer, offset).set(bytes);
  }
  public get f3() { return this.buffer.dv.getUint8(this.offset+16); }
  public set f3(value: number) { this.buffer.dv.setUint8(this.offset+16,value); }
  public get f4() { return this.buffer.dv.getUint16(this.offset+18,true); }
  public set f4(value: number) { this.buffer.dv.setUint16(this.offset+18,value,true); }
  public get f5() { return this.buffer.dv.getUint32(this.offset+20,true); }
  public set f5(value: number) { this.buffer.dv.setUint32(this.offset+20,value,true); }
  public f6(elements_size: number): void { 
    NPRPC.Flat._alloc(this.buffer, this.offset + 24, elements_size, 12, 4);
  }
  public f6_d() { return new NPRPC.Flat.Vector_Direct2<Flat_benchmark.Data1_Direct>(this.buffer, this.offset + 24, 12, Flat_benchmark.Data1_Direct); }
}
} // namespace Flat 
export class Benchmark extends NPRPC.ObjectProxy {
  public static get servant_t(): new() => _IBenchmark_Servant {
    return _IBenchmark_Servant;
  }


  public async rpc(): Promise<Data> {
    let interface_idx = (arguments.length == 0 ? 0 : arguments[arguments.length - 1]);
    let buf = NPRPC.FlatBuffer.create();
    buf.prepare(32);
    buf.commit(32);
    buf.write_msg_id(NPRPC.impl.MessageId.FunctionCall);
    buf.write_msg_type(NPRPC.impl.MessageType.Request);
    let __ch = new NPRPC.impl.Flat_nprpc_base.CallHeader_Direct(buf, 16);
    __ch.object_id = this.data.object_id;
    __ch.poa_idx = this.data.poa_idx;
    __ch.interface_idx = interface_idx;
    __ch.function_idx = 0;
    buf.write_len(buf.size - 4);
    await NPRPC.rpc.call(
      this.endpoint(), buf, this.timeout
    );
    let std_reply = NPRPC.handle_standart_reply(buf);
    if (std_reply != -1) {
      console.log("received an unusual reply for function with output arguments");
      throw new NPRPC.Exception("Unknown Error");
    }
  let out = new Flat_benchmark.benchmark_M1_Direct(buf, 16);
  let __ret_value: Data;
  (__ret_value as any) = new Object();
  __ret_value.f1 = out._1.f1;
  __ret_value.f2 = out._1.f2;
  __ret_value.f3 = out._1.f3;
  __ret_value.f4 = out._1.f4;
  __ret_value.f5 = out._1.f5;
  {
  let vv = out._1.f6_d(), index_0 = 0;
  (__ret_value.f6 as Array<any>) = new Array<any>(vv.elements_size)
  for (let e of vv) {
    (__ret_value.f6[index_0] as any) = new Object();
  (__ret_value.f6[index_0] as any) = new Object();
  __ret_value.f6[index_0].f1 = e.f1;
  __ret_value.f6[index_0].f2 = e.f2;
    ++index_0;
  }
  }
  return __ret_value;
}

  public async json(): Promise<string> {
    let interface_idx = (arguments.length == 0 ? 0 : arguments[arguments.length - 1]);
    let buf = NPRPC.FlatBuffer.create();
    buf.prepare(32);
    buf.commit(32);
    buf.write_msg_id(NPRPC.impl.MessageId.FunctionCall);
    buf.write_msg_type(NPRPC.impl.MessageType.Request);
    let __ch = new NPRPC.impl.Flat_nprpc_base.CallHeader_Direct(buf, 16);
    __ch.object_id = this.data.object_id;
    __ch.poa_idx = this.data.poa_idx;
    __ch.interface_idx = interface_idx;
    __ch.function_idx = 1;
    buf.write_len(buf.size - 4);
    await NPRPC.rpc.call(
      this.endpoint(), buf, this.timeout
    );
    let std_reply = NPRPC.handle_standart_reply(buf);
    if (std_reply != -1) {
      console.log("received an unusual reply for function with output arguments");
      throw new NPRPC.Exception("Unknown Error");
    }
  let out = new Flat_benchmark.benchmark_M2_Direct(buf, 16);
  let __ret_value: string;
  __ret_value = out._1;
  return __ret_value;
}

};

export interface IBenchmark_Servant
{
  rpc(): Data;
  json(): string;
}

export class _IBenchmark_Servant extends NPRPC.ObjectServant {
  public static _get_class(): string { return "benchmark/benchmark.Benchmark"; }
  public readonly get_class = () => { return _IBenchmark_Servant._get_class(); }
  public readonly dispatch = (buf: NPRPC.FlatBuffer, remote_endpoint: NPRPC.EndPoint, from_parent: boolean) => {
    _IBenchmark_Servant._dispatch(this, buf, remote_endpoint, from_parent);
  }
  static _dispatch(obj: _IBenchmark_Servant, buf: NPRPC.FlatBuffer, remote_endpoint: NPRPC.EndPoint, from_parent: boolean): void {
  let __ch = new NPRPC.impl.Flat_nprpc_base.CallHeader_Direct(buf, 16);
  switch(__ch.function_idx) {
    case 0: {
      let obuf = buf;
      obuf.consume(obuf.size);
      obuf.prepare(176);
      obuf.commit(48);
      let oa = new Flat_benchmark.benchmark_M1_Direct(obuf,16);
let __ret_val: Data;
      __ret_val = (obj as any).rpc();
  oa._1.f1 = __ret_val.f1;
  oa._1.f2 = __ret_val.f2;
  oa._1.f3 = __ret_val.f3;
  oa._1.f4 = __ret_val.f4;
  oa._1.f5 = __ret_val.f5;
  oa._1.f6(__ret_val.f6.length);
  {
  let vv = oa._1.f6_d(), index = 0;
  for (let e of vv) {
          e.f1 = __ret_val.f6[index].f1;
      e.f2 = __ret_val.f6[index].f2;
    ++index;
  }
  }
      obuf.write_len(obuf.size - 4);
      obuf.write_msg_id(NPRPC.impl.MessageId.BlockResponse);
      obuf.write_msg_type(NPRPC.impl.MessageType.Answer);
      break;
    }
    case 1: {
      let obuf = buf;
      obuf.consume(obuf.size);
      obuf.prepare(152);
      obuf.commit(24);
      let oa = new Flat_benchmark.benchmark_M2_Direct(obuf,16);
let __ret_val: string;
      __ret_val = (obj as any).json();
  oa._1 = __ret_val;
      obuf.write_len(obuf.size - 4);
      obuf.write_msg_id(NPRPC.impl.MessageId.BlockResponse);
      obuf.write_msg_type(NPRPC.impl.MessageType.Answer);
      break;
    }
    default:
      NPRPC.make_simple_answer(buf, NPRPC.impl.MessageId.Error_UnknownFunctionIdx);
  }
  }
}

export interface benchmark_M1 {
  _1: Data;
}

export namespace Flat_benchmark {
export class benchmark_M1_Direct extends NPRPC.Flat.Flat {
  public get _1() { return new Data_Direct(this.buffer, this.offset + 0); }
}
} // namespace Flat 
export interface benchmark_M2 {
  _1: string;
}

export namespace Flat_benchmark {
export class benchmark_M2_Direct extends NPRPC.Flat.Flat {
  public get _1() {
    const offset = this.offset + 0;
    const n = this.buffer.dv.getUint32(offset + 4, true);
    return n > 0 ? u8dec.decode(new DataView(this.buffer.array_buffer, offset + this.buffer.dv.getUint32(offset, true), n)) : ""
  }
  public set _1(str: string) {
    const bytes = u8enc.encode(str);
    const offset = NPRPC.Flat._alloc(this.buffer, this.offset + 0, bytes.length, 1, 1);
    new Uint8Array(this.buffer.array_buffer, offset).set(bytes);
  }
}
} // namespace Flat 

File: nprpc_benchmark/benchmark.cpp
#include "benchmark.hpp"
#include <nprpc/impl/nprpc_impl.hpp>

void benchmark_throw_exception(::nprpc::flat_buffer& buf);

namespace {
struct benchmark_M1 {
  benchmark::flat::Data _1;
};

class benchmark_M1_Direct {
  ::nprpc::flat_buffer& buffer_;
  const std::uint32_t offset_;

  auto& base() noexcept { return *reinterpret_cast<benchmark_M1*>(reinterpret_cast<std::byte*>(buffer_.data().data()) + offset_); }
  auto const& base() const noexcept { return *reinterpret_cast<const benchmark_M1*>(reinterpret_cast<const std::byte*>(buffer_.data().data()) + offset_); }
public:
  uint32_t offset() const noexcept { return offset_; }
  void* __data() noexcept { return (void*)&base(); }
  benchmark_M1_Direct(::nprpc::flat_buffer& buffer, std::uint32_t offset)
    : buffer_(buffer)
    , offset_(offset)
  {
  }
  auto _1() noexcept { return benchmark::flat::Data_Direct(buffer_, offset_ + offsetof(benchmark_M1, _1)); }
};

struct benchmark_M2 {
  ::nprpc::flat::String _1;
};

class benchmark_M2_Direct {
  ::nprpc::flat_buffer& buffer_;
  const std::uint32_t offset_;

  auto& base() noexcept { return *reinterpret_cast<benchmark_M2*>(reinterpret_cast<std::byte*>(buffer_.data().data()) + offset_); }
  auto const& base() const noexcept { return *reinterpret_cast<const benchmark_M2*>(reinterpret_cast<const std::byte*>(buffer_.data().data()) + offset_); }
public:
  uint32_t offset() const noexcept { return offset_; }
  void* __data() noexcept { return (void*)&base(); }
  benchmark_M2_Direct(::nprpc::flat_buffer& buffer, std::uint32_t offset)
    : buffer_(buffer)
    , offset_(offset)
  {
  }
  void _1(const char* str) { new (&base()._1) ::nprpc::flat::String(buffer_, str); }
  void _1(const std::string& str) { new (&base()._1) ::nprpc::flat::String(buffer_, str); }
  auto _1() noexcept { return (::nprpc::flat::Span<char>)base()._1; }
  auto _1() const noexcept { return (::nprpc::flat::Span<const char>)base()._1; }
  auto _1_d() noexcept {     return ::nprpc::flat::String_Direct1(buffer_, offset_ + offsetof(benchmark_M2, _1));  }
};


} // 

namespace benchmark { 
Data benchmark::Benchmark::rpc() {
  ::nprpc::flat_buffer buf;
  {
    auto mb = buf.prepare(32);
    buf.commit(32);
    static_cast<::nprpc::impl::Header*>(mb.data())->msg_id = ::nprpc::impl::MessageId::FunctionCall;
    static_cast<::nprpc::impl::Header*>(mb.data())->msg_type = ::nprpc::impl::MessageType::Request;
  }
  ::nprpc::impl::flat::CallHeader_Direct __ch(buf, sizeof(::nprpc::impl::Header));
  __ch.object_id() = this->_data().object_id;
  __ch.poa_idx() = this->_data().poa_idx;
  __ch.interface_idx() = interface_idx_;
  __ch.function_idx() = 0;
  static_cast<::nprpc::impl::Header*>(buf.data().data())->size = static_cast<uint32_t>(buf.size() - 4);
  ::nprpc::impl::g_orb->call(this->get_endpoint(), buf, this->get_timeout());
  auto std_reply = nprpc::impl::handle_standart_reply(buf);
  benchmark_M1_Direct out(buf, sizeof(::nprpc::impl::Header));
    Data __ret_value;
    __ret_value.f1 = (std::string_view)out._1().f1();
    __ret_value.f2 = (std::string_view)out._1().f2();
    __ret_value.f3 = out._1().f3();
    __ret_value.f4 = out._1().f4();
    __ret_value.f5 = out._1().f5();
    {
      auto span = out._1().f6();
      __ret_value.f6.resize(span.size());
      auto it3 = std::begin(__ret_value.f6);
      for (auto e : span) {
        (*it3).f1 = (std::string_view)e.f1();
        (*it3).f2 = e.f2();
        ++it3;
      }
    }
  return __ret_value;
}

std::string benchmark::Benchmark::json() {
  ::nprpc::flat_buffer buf;
  {
    auto mb = buf.prepare(32);
    buf.commit(32);
    static_cast<::nprpc::impl::Header*>(mb.data())->msg_id = ::nprpc::impl::MessageId::FunctionCall;
    static_cast<::nprpc::impl::Header*>(mb.data())->msg_type = ::nprpc::impl::MessageType::Request;
  }
  ::nprpc::impl::flat::CallHeader_Direct __ch(buf, sizeof(::nprpc::impl::Header));
  __ch.object_id() = this->_data().object_id;
  __ch.poa_idx() = this->_data().poa_idx;
  __ch.interface_idx() = interface_idx_;
  __ch.function_idx() = 1;
  static_cast<::nprpc::impl::Header*>(buf.data().data())->size = static_cast<uint32_t>(buf.size() - 4);
  ::nprpc::impl::g_orb->call(this->get_endpoint(), buf, this->get_timeout());
  auto std_reply = nprpc::impl::handle_standart_reply(buf);
  benchmark_M2_Direct out(buf, sizeof(::nprpc::impl::Header));
    std::string __ret_value;
    __ret_value = (std::string_view)out._1();
  return __ret_value;
}

void benchmark::IBenchmark_Servant::dispatch(nprpc::Buffers& bufs, nprpc::EndPoint remote_endpoint, bool from_parent, nprpc::ReferenceList& ref_list) {
  nprpc::impl::flat::CallHeader_Direct __ch(bufs(), sizeof(::nprpc::impl::Header));
  switch(__ch.function_idx()) {
    case 0: {
      auto& obuf = bufs.flip();
      obuf.consume(obuf.size());
      obuf.prepare(176);
      obuf.commit(48);
      benchmark_M1_Direct oa(obuf,16);
      Data __ret_val;
      __ret_val = rpc();
      oa._1().f1(__ret_val.f1);
      oa._1().f2(__ret_val.f2);
      oa._1().f3() = __ret_val.f3;
      oa._1().f4() = __ret_val.f4;
      oa._1().f5() = __ret_val.f5;
      oa._1().f6(static_cast<uint32_t>(__ret_val.f6.size()));
      {
        auto span = oa._1().f6();
        auto it = __ret_val.f6.begin();
        for (auto e : span) {
          auto __ptr = ::nprpc::make_wrapper1(*it);
            e.f1(__ptr->f1);
            e.f2() = __ptr->f2;
          ++it;
        }
      }
      static_cast<::nprpc::impl::Header*>(obuf.data().data())->size = static_cast<uint32_t>(obuf.size() - 4);
      static_cast<::nprpc::impl::Header*>(obuf.data().data())->msg_id = ::nprpc::impl::MessageId::BlockResponse;
      static_cast<::nprpc::impl::Header*>(obuf.data().data())->msg_type = ::nprpc::impl::MessageType::Answer;
      break;
    }
    case 1: {
      auto& obuf = bufs.flip();
      obuf.consume(obuf.size());
      obuf.prepare(152);
      obuf.commit(24);
      benchmark_M2_Direct oa(obuf,16);
      std::string __ret_val;
      __ret_val = json();
      oa._1(__ret_val);
      static_cast<::nprpc::impl::Header*>(obuf.data().data())->size = static_cast<uint32_t>(obuf.size() - 4);
      static_cast<::nprpc::impl::Header*>(obuf.data().data())->msg_id = ::nprpc::impl::MessageId::BlockResponse;
      static_cast<::nprpc::impl::Header*>(obuf.data().data())->msg_type = ::nprpc::impl::MessageType::Answer;
      break;
    }
    default:
      nprpc::impl::make_simple_answer(bufs(), nprpc::impl::MessageId::Error_UnknownFunctionIdx);
  }
}

} // namespace benchmark

